### PR TITLE
a word about crontasks

### PR DESCRIPTION
**Pull request by btry**

### CODE REVIEW

**Comment by trasher** [comment]
Hello,

I cannot activate build for the branch since this is on your own repository.

BTW, this is stille marked as WIP, but you requested a review... Which is wrong? ;)

**Comment by btry** [comment]
Hi

I removed the WIP label. If build is not allowed from the fork, then I can push the branch here.

**Comment by btry** [comment]
PR updated with the same content as https://github.com/glpi-project/docdev/tree/feature/crontask to ease review.

File: source/devapi/crontasks.rst [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+^^^^^
+
+Provide a scheduler for background tasks used by GLPI and its plugins.
+
+Implementation overview
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The scheduler wakes up thanks to user activity or from the operating system's scheduler. On each execution, it executes a limited number of automatic actions.
+
+The entry point of automatic actions is the file front/cron.php. It is called from the command line or from a web page when a user browses in GLPI.
```
Comment (orthagh) [inline source/devapi/crontasks.rst]: a word on invisible image launch ?

File: source/devapi/crontasks.rst [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+The automatic actions are defined by the class CronTask. GLPI defines a lot of them for its own needs. They are created in the installation or upgrade process.
+
+Implementation
+^^^^^^^^^^^^^^
+
+Automatic actions are implemented in static methods of itemtypes. The itemtype containing the automatic action should be an itemtype related to the action. However some automatic actions are related to themselves or cannot be related to an itemtype. Such automatic actions are implemented in CronTask.
```
Comment (orthagh) [inline source/devapi/crontasks.rst]: This part is pretty obscure/

I suggest:

```
Automatic actions could be related to an itemtype  and the implemention is defined in its class or haven't any itemtype relation and are implemented directly into CronTask class
-

File: source/devapi/crontasks.rst [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+When GLPI shows a list of automatic actions, it shows a short description for each item. The description is gathered in the static method ``cronInfo()`` in the itemtype containing the reltaed automatic action.
```
Comment (orthagh) [inline source/devapi/crontasks.rst]: typo: reltaed
BTW, could be shorter "...in the itemtype class"

File: source/devapi/crontasks.rst [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            }
+         }
+         return $cron_status;
+      }
+
+      // ...
+
+   }
+
+If the argument $task is a CronTask object, the automatic action must increment the quantity of actions done. In this example, each email actually sent increments the volume by 1.
```
Comment (orthagh) [inline source/devapi/crontasks.rst]: the automatic action must increment the quantity of actions done 
-> 
the method|process must increment the volume of actions done

File: source/devapi/crontasks.rst [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+Register an automatic actions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Automatic actions are defined in the empty schema located in ``install/mysql/``. Use  the existing sql queries creating rows in the table ``glpi_crontasks`` as template for a new automatic action.
+
+To handle upgrade from a previous version, the new automatic actions must be added in the appropriate update file in ``install/``.
```
Comment (orthagh) [inline source/devapi/crontasks.rst]: install/update_xx_to_yy.php

File: source/plugins/crontasks.rst [hunk 6]
```diff
@@ -0,0 +1,42 @@
+Automatic actions
+-----------------
+
+Goals
+^^^^^
+
+Plugins may need to run automatic actions in backgroupd, or at regular interval. GLPI provides a task scheduler for itself and its plugins.
```
Comment (orthagh) [inline source/plugins/crontasks.rst]: typo: backgroupd

File: source/plugins/crontasks.rst [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+Implement an automatic action
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A plugin must implement its automatic action the same way as GLPI does, except the method is located in a plugin's itemtype. See :doc:`crontasks <../devapi/crontasks>`.
+
+Register an automatic action
```
Comment (orthagh) [inline source/plugins/crontasks.rst]: As a lot of text is duplicated from core content (like arguments), maybe a single link with a sort introduction and differences may be better.

File: source/devapi/crontasks.rst (hunk 5 above)
Comment (btry) [inline source/devapi/crontasks.rst]: I think we need a dedicated page about the update process, to cover the update_xx_to_yy. Anyway I update as you suggested.

File: source/plugins/crontasks.rst (hunk 7 above)
Comment (btry) [inline source/plugins/crontasks.rst]: I redirected to reader to devapi.

File: source/devapi/crontasks.rst [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+Goals
+^^^^^
+
+Provide a scheduler for background tasks used by GLPI and its plugins. 
+
+Implementation overview
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The entry point of automatic actions is the file front/cron.php. On each execution, it executes a limited number of automatic actions.
```
Comment (trasher) [inline source/devapi/crontasks.rst]:
```
``front/cron.php``
```

File: source/devapi/crontasks.rst [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+There are two ways to wake up the scheduler :
+ - when a user browses in GLPI (the internal mode)
+ - when the operating system's scheduler calls ``front/cron.php`` (the external mode)
+
+When GLPI generates an HTML page for a browser, it adds an invisible image generated by ``front/cron.php``. This way, the automatic action runs in a separate process and does not imapct the user.
+
+The automatic actions are defined by the class CronTask. GLPI defines a lot of them for its own needs. They are created in the installation or upgrade process.
```
Comment (trasher) [inline source/devapi/crontasks.rst]: Link class to apidoc

File: source/devapi/crontasks.rst [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+Automatic actions could be related to an itemtype and the implemention is defined in its class or haven't any itemtype relation and are implemented directly into CronTask class.
+
+When GLPI shows a list of automatic actions, it shows a short description for each item. The description is gathered in the static method ``cronInfo()`` of the itemtype.
+
+.. Note::
+
+   An itemtype may contain several automatic actions.
+
+Example of implemtation from the class ``QueuedMail``:
```
Comment (trasher) [inline source/devapi/crontasks.rst]: implemtation typo
Maybe link to QueudMail class apidoc?

File: source/devapi/crontasks.rst [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            }
+         }
+         return $cron_status;
+      }
+
+      // ...
+
+   }
+
+If the argument $task is a CronTask object, the method must increment the quantity of actions done. In this example, each email actually sent increments the volume by 1.
```
Comment (trasher) [inline source/devapi/crontasks.rst]:
```
``$task``
```

File: source/devapi/crontasks.rst [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+   // Register an automatic action
+   CronTask::register('QueuedMail', 'QueuedMail', HOUR_TIMESTAMP,
+         array(
+         'comment'   => '',
+         'mode'      => CronTask::MODE_EXTERNAL
+   ));
+
+The ``register`` method takes four arguments:
+
+* `itemtype`: a `string` containing an itemtype name containing the automatic action implementation
```
Comment (trasher) [inline source/devapi/crontasks.rst]: Why single `? Should double

File: source/devapi/crontasks.rst [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+* `name`: a `string` containing the name of the automatic action
+* `frequency` the period of time between two executions in seconds (see inc/define.php for convenient constants)
```
Comment (trasher) [inline source/devapi/crontasks.rst]:
```
``inc/define.php``
```

File: source/devapi/crontasks.rst (hunk 10 above)
Comment (btry) [inline source/devapi/crontasks.rst]: Ok for the typo, of course.

I not sure about linking to QueudMail : the page is about CronTask, and the reader should not read too much about QueuedMail from here. It is used as an example only.

Comment (trasher) [inline source/devapi/crontasks.rst]: Ok :)

**Comment by trasher** [comment]
Merged; see 9db29deea089dd3f04456e719f57d5109b574a01

**Pull request closed without merging**